File: integration-tests/src/test/java/dev/capylang/test/LocalTypesAndDataTest.java
package dev.capylang.test;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class LocalTypesAndDataTest {
    @Test
    void localTypesAndDataInsideFunctionShouldWork() {
        assertThat(LocalTypesAndData.fooMe("foo")).isEqualTo("xyz");
        assertThat(LocalTypesAndData.fooMe("boo")).isEqualTo("zyx");
        assertThat(LocalTypesAndData.fooMe("john")).isEqualTo("john");
    }

    @Test
    void localDataShouldResolveImportedResultTypeWithoutFullyQualifiedName() {
        assertThat(LocalTypesAndData.localDataWithImportedResult(7)).isEqualTo("ok:7");
    }

    @Test
    void localFunctionShouldResolveLocalDataTypeInSignature() {
        assertThat(LocalTypesAndData.localDataUsedInLocalFunction(7)).isEqualTo("value:7");
    }

    @Test
    void localFunctionShouldResolveLocalGenericDataTypeInsideNestedReturnType() {
        assertThat(LocalTypesAndData.localGenericDataUsedInLocalFunctionReturnType(7)).isEqualTo("value:7");
    }

    @Test
    void localGenericDataFieldAccessShouldRemainVisibleBeforeIndexingInsidePipeLambda() {
        assertThat(LocalTypesAndData.localGenericDataFieldAccessFollowedByIndex("abc")).isEqualTo("a");
    }

    @Test
    void genericOptionConstructorPatternShouldBindConcreteType() {
        assertThat(LocalTypesAndData.localGenericOptionConstructorPatternBinding(7)).isEqualTo(77);
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/LocalTypesAndData.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('LocalTypesAndData', () => {
    const localTypes = generatedModule('dev/capylang/test/LocalTypesAndData.js');
    assert.equal(localTypes.fooMe('foo'), 'xyz');
    assert.equal(localTypes.localDataWithImportedResult(7), 'ok:7');
    assert.equal(localTypes.localSingleUsedInLocalTypeAndFunction(true), 'stop');
    assert.equal(localTypes.localGenericDataFieldAccessFollowedByIndex('abc'), 'a');
    assert.equal(localTypes.localGenericOptionConstructorPatternBinding(7), 77);
});
